door-runtime/src/jsMain/kotlin/com/ustadmobile/door/DoorDatabase.kt

```diff
@@ -1,35 +1,71 @@
 package com.ustadmobile.door
 
-import com.ustadmobile.door.jdbc.Connection
-import kotlinx.coroutines.CompletableDeferred
-import kotlinx.coroutines.GlobalScope
-import kotlinx.coroutines.Runnable
-import kotlinx.coroutines.launch
-import wrappers.SQLiteDatasourceJs
+import com.ustadmobile.door.ext.useConnection
+import com.ustadmobile.door.jdbc.ResultSet
+import com.ustadmobile.door.jdbc.SQLException
+import kotlinx.coroutines.*
+import wrappers.*
+import kotlin.jvm.Volatile
 
 actual abstract class DoorDatabase actual constructor(): DoorDatabaseCommon() {
 
-    override val jdbcDbType: Int
-        get() = DoorDbType.SQLITE
+    override val jdbcDbType: Int = DoorDbType.SQLITE
 
-    override val jdbcArraySupported: Boolean
-        get() = false
-
-    internal lateinit var webWorkerPath: String
-
-    val initCompletable = CompletableDeferred<Boolean>()
-
-    private suspend fun awaitReady() {
-        initCompletable.await()
-    }
+    @Volatile
+    override val jdbcArraySupported: Boolean = false
 
     actual abstract fun clearAllTables()
 
     actual override fun runInTransaction(runnable: Runnable) {
         super.runInTransaction(runnable)
     }
 
+    override val tableNames: List<String>
+        get() = throw Exception("This can't be used in JS, only on JVM use getTableNamesAsync")
+
+    suspend fun getTableNamesAsync(): List<String> {
+        val delegatedDatabaseVal = sourceDatabase
+        return if(delegatedDatabaseVal != null) {
+            delegatedDatabaseVal.getTableNamesAsync()
+        } else {
+            var con = null as SQLiteConnectionJs?
+            val tableNamesList = mutableListOf<String>()
+            val tableResult: ResultSet?
+            try {
+                con = openConnection() as SQLiteConnectionJs
+                val metadata = con.getMetaData() as SQLiteDatabaseMetadataJs
+                tableResult = metadata.getTablesAsync(null, null, "%", arrayOf("TABLE"))
+                while(tableResult.next()) {
+                    tableResult.getString("TABLE_NAME")?.also {
+                        tableNamesList.add(it)
+                    }
+                }
+            }finally {
+                con?.close()
+            }
+
+            tableNamesList.toList()
+        }
+    }
+
+    suspend fun execSQLBatchAsync(vararg sqlStatements: String){
+        var connection: SQLiteConnectionJs? = null
+        try {
+            connection = openConnection() as SQLiteConnectionJs
+            connection.setAutoCommit(false)
+            val statement = SQLitePreparedStatementJs(connection,sqlStatements.joinToString(";"))
+            val result = statement.executeUpdateAsync()
+            if(result != -1){
```

Comment:
I think this if result != -1 check should be removed. If something goes wrong, an exception should be thrown.
This should be using the normal Statement rather than PreparedStatement. PreparedStatement tells the database to parse and hold onto a statement (as I understand the SQL is normally compiled into object code at this stage). This improves performance. 

That isn't really being done on Javascript right now, but it might be needed in future for performance. 

The query being executed here is not 'prepared' or to be cached. It is a set of one off statements to execute. So it should use the normal Statement.